File: app/android/java/MoviesApp/app/src/main/java/com/example/moviesapp/view/BaseViewImpl.java
package com.example.moviesapp.view;

import androidx.databinding.ViewDataBinding;

/**
 * Base view implementation.
 *
 * @param <T> View binding
 */
public abstract class BaseViewImpl<T extends ViewDataBinding>
        implements ScreenView<T> {

    private T binding;

    @Override
    public void applyBinding(T binding) {
        this.binding = binding;
        setBinding(binding);
    }

    @Override
    public void resetBinding() {
        clearBinding();
        binding = null;
    }

    /**
     * Returns binding.
     */
    protected final T binding() {
        return binding;
    }

    /**
     * Sets binding.
     *
     * @param binding view binding
     */
    protected abstract void setBinding(T binding);

    /**
     * Clears binding.
     */
    protected abstract void clearBinding();
}

File: app/android/kotlin/MoviesApp/app/src/main/java/com/example/moviesapp/view/BaseViewImpl.kt
package com.example.moviesapp.view

import androidx.databinding.ViewDataBinding

/**
 * Base view implementation.
 *
 * @param <T> View binding
 */
abstract class BaseViewImpl<T : ViewDataBinding> : ScreenView<T> {

    private var binding: T? = null

    override fun applyBinding(binding: T) {
        this.binding = binding
        setBinding(binding)
    }

    override fun resetBinding() {
        clearBinding()
        binding = null
    }

    /**
     * Returns binding.
     */
    protected fun binding(): T? {
        return binding
    }

    /**
     * Sets binding.
     *
     * @param binding view binding
     */
    protected abstract fun setBinding(binding: T)

    /**
     * Clears binding.
     */
    protected abstract fun clearBinding()
}